### PR TITLE
fix(model): load ModelOptions lazily

### PR DESCRIPTION
**Pull request by sauyon**
Also add a less opinionated base `ModelOptions`; that's not really necessary and I'm not actually sure I want it to exist. @ssheng thoughts?

### CODE REVIEW

File: bentoml/_internal/models/model.py [hunk 1]
```diff
@@ -56,19 +58,25 @@ class ModelSignatureDict(t.TypedDict, total=False):
 CUSTOM_OBJECTS_FILENAME = "custom_objects.pkl"
 
 
-@attr.define
 class ModelOptions:
+    def with_options(self, **kwargs: t.Any) -> ModelOptions:
+        if len(kwargs) > 0:
+            raise InvalidArgument("This model does not support any options.")
```
Comment (ssheng) [inline bentoml/_internal/models/model.py:64]: More accurately this "framework" does not support any options. Can we pass the`info.module` field into `ModelOptions` and include the module name in the exception message?

File: bentoml/_internal/models/model.py [hunk 2]
```diff
@@ -532,6 +546,36 @@ def with_options(self, **kwargs: t.Any) -> ModelInfo:
             creation_time=self.creation_time,
         )
 
+    # cached_property doesn't support __slots__ classes
+    @property
+    def imported_module(self) -> ModuleType:
```
Comment (ssheng) [inline bentoml/_internal/models/model.py:551]: May be we can get rid of `self._cached_module` if we use a `@lru_cache(maxsize=1)` decorator here.

**Review by ssheng — COMMENTED** [review]
Thanks, Sauyon. I think this is a good and clean solution.

File: bentoml/_internal/models/model.py [hunk 3]
```diff
@@ -56,19 +58,25 @@ class ModelSignatureDict(t.TypedDict, total=False):
...
+        if len(kwargs) > 0:
+            raise InvalidArgument("This model does not support any options.")
+        return self
+
+    def to_dict(self) -> dict[str, t.Any]:
+        return {}
+
+
+@attr.define
+class AttrsModelOptions(ModelOptions):
```
Comment (parano) [inline bentoml/_internal/models/model.py:72]: add this to `bentoml.models`?

File: bentoml/_internal/models/model.py (hunk 2 above)
Comment (aarnphm) [reply bentoml/_internal/models/model.py:551]: i don't think that would work?

Comment (sauyon) [reply bentoml/_internal/models/model.py:551]: I would think `lru_cache` also uses the dict, but will test this one out.

File: bentoml/_internal/frameworks/FRAMEWORK_TEMPLATE_PY [hunk 4]
```diff
@@ -46,9 +46,12 @@ def get(tag_like: str | Tag) -> bentoml.Model:
...
 
 
-@attr.define(frozen=True)
 class FrameworkOptions(ModelOptions):
-    pass
+    def with_options(self, **kwargs: t.Any) -> FrameworkOptions:
+        ...
+
+    def to_dict(self) -> dict[str, t.Any]:
+        ...
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply bentoml/_internal/frameworks/FRAMEWORK_TEMPLATE_PY:54]: Exporting `AttrsModelOptions` to `bentoml.models` would solve this issue

File: bentoml/_internal/models/model.py (hunk 1 above)
Comment (sauyon) [reply bentoml/_internal/models/model.py:64]: We probably could, let's do that.

File: bentoml/_internal/models/model.py (hunk 2 above)
Comment (sauyon) [reply bentoml/_internal/models/model.py:551]: Hm, after reading up on this `lru_cache` might not do exactly what we want; it'll cache externally over the `self` parameter without modifying the actual object so we'd have to either let it be unbounded, which would be a leak, or only cache a certain number.

**Review by aarnphm — APPROVED** [review]

**Pull request merged**